src/keys.rs

```diff
@@ -841,6 +865,20 @@ impl EphemeralPublicKey {
     }
 }
 
+/// A Sapling ephemeral public key that has been precomputed for trial decryption.
+#[derive(Clone, Debug)]
+pub struct PreparedEphemeralPublicKey(PreparedNonIdentityBase);
```

Comment:
This looks odd: in Sapling non-identity points have no special status. Non-small order points do, and prime-order points pdo.

Reply:
I see the issue: $\mathsf{GroupHash}^{\mathbb{J}^{(r)\ast}}$ returns a prime-order point but is fallible, and unlike in Orchard, a $\bot$ return is not replaced with another point.

Reply:
Sorry I must not be quite awake yet. This is the orchard crate; the comment should say Orchard.